### PR TITLE
feat(platform): Added screen for CREATE NEW PROJECT

### LINKED ISSUE #244: PLATFORM: Add create new project
## Description

Implement a “create new project” functionality in the dashboard of the platform. The created new project should be immediately updated in the project dashboard. Some part of create new project is done, which can find in [`apps/platform/src/app/(main)/page.tsx`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/platform/src/app/(main)/page.tsx). 

## Scope of work

You have to implement 2 features from create new project. 
1. Add environment for the project, by default there is an environment named `default`. In the array of object environment, there should be only one default workspace. And user can add as many environments as they want.
2. After sending the environment request. The project dashboard should auto refresh to show the recently added project.

### Postman URL
- [Create project](https://www.postman.com/blue-crescent-581920/workspace/keyshade/api/a31bdb66-69e3-469b-afb4-f2051385e634/request/32733901-299c8b25-c346-4610-b3a2-dd043728da3c?collection=32733901-fae1f431-06fa-4425-aef4-65b28e9e9e71&version=63ff7d1a-5155-4cf3-ae4f-75c552aef059)

### Type of new project and environment

https://github.com/keyshade-xyz/keyshade/blob/f3a36326b4f5c945fb2725620ff92ab31e44e053/apps/platform/src/types/index.ts#L43-L54

### Dialog for create new project

https://github.com/keyshade-xyz/keyshade/blob/f3a36326b4f5c945fb2725620ff92ab31e44e053/apps/platform/src/app/(main)/page.tsx#L116-L187

### PR DESCRIPTION
**Pull request by poswalsameer**
## Description
Added screen for `Create New Project` by creating the UI and adding all the required functions to create a new project on button click.

Fixes #244

## Dependencies
No dependencies.

## Mentions
@rajdip-b @kriptonian1 

## Screenshots of relevant screens
Screenshot (521)
Screenshot (246)

## Developer's checklist

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [x] I have added relevant screenshots in my PR
- [x] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

___

### **PR Type**
Enhancement, Dependencies

___

### **Description**
- Implemented the "create new project" feature in the platform dashboard, allowing users to create projects with specified environments and access levels.
- Added a new select component using Radix UI to enhance the UI for selecting access levels.
- Updated type definitions to include `accessLevel` for projects.
- Added `@radix-ui/react-select` as a new dependency to support the select component.
- Updated TypeScript configuration to support new JSX and path settings.
- Updated the lockfile to reflect changes in dependencies.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>page.tsx</strong><dd><code>Implement "Create New Project" Feature in Dashboard</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/src/app/(main)/page.tsx

<li>Implemented the "create new project" functionality.<br> <li> Added UI components for project creation.<br> <li> Integrated project creation with backend API.<br> <li> Updated project state management.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-d6c7ffe7a9ea578ea4b0a5587f2759a7a1eac9a7c7d6dfae25ddd0bae48ea8e2">+173/-45</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>select.tsx</strong><dd><code>Add Select Component Using Radix UI</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/src/components/ui/select.tsx

<li>Added new UI select component.<br> <li> Utilized Radix UI for select functionality.<br> <li> Styled select component for consistency.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-aeb5c4d8a30ec714f81155cc535121efc75ba0d8dc97356151b198842a64afa0">+158/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Update Type Definitions for Project Creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/src/types/index.ts

<li>Added <code>accessLevel</code> to <code>NewProject</code> type.<br> <li> Updated type definitions for project creation.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-959c8209f1c5cf5002c053941c00735d40f5352a55dacd687067f06ad751c0e0">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Add Radix UI Select Dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/package.json

- Added `@radix-ui/react-select` dependency.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-ccbe95f6e6aa4e3266a5e9cf1bcfee8f323da5b0be41f8cab6e3f8ab98d0ddcd">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pnpm-lock.yaml</strong><dd><code>Update Lockfile for New Dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pnpm-lock.yaml

- Updated lockfile with new dependencies.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bb">+175/-301</a></td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>tsconfig.json</strong><dd><code>Update TypeScript Configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/tsconfig.json

- Updated TypeScript configuration for JSX and paths.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/540/files#diff-494c9c390ce40cab77d1a6ea5ca371d57c8d3c1ca9b868104e79a6714d64bfce">+22/-6</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: apps/platform/src/app/(main)/page.tsx [hunk 1]
```diff
@@ -77,20 +77,23 @@ export default function Index(): JSX.Element {
...
-        isForked: response.data.isForked,
-        forkedFromId: response.data.forkedFromId
+        id: data.id,
+        name: data.name,
+        description: data.description,
+        createdAt: data.createdAt,
+        updatedAt: data.updatedAt,
+        storePrivateKey: data.storePrivateKey,
+        isDisabled: data.isDisabled,
+        accessLevel: data.accessLevel as "GLOBAL" | "INTERNAL" | "PRIVATE",
```
Comment (rajdip-b) [inline apps/platform/src/app/(main)/page.tsx:92]: Access level is already inferred as these types.

File: apps/platform/src/app/(main)/page.tsx [hunk 2]
```diff
@@ -53,16 +53,16 @@ export default function Index(): JSX.Element {
     environments: [
       {
         name: '',
-        description: '',
-        isDefault: false
+        projectId: Date.now().toString(),
```
Comment (rajdip-b) [inline apps/platform/src/app/(main)/page.tsx:56]: We can just set it to ''. Either ways, this won't be sent to the backend

File: apps/platform/src/types/index.ts [hunk 3]
```diff
@@ -42,8 +42,12 @@ export const zProject = z.object({
 
 export const zEnvironment = z.object({
   name: z.string(),
-  description: z.string().nullable(),
-  isDefault: z.boolean().optional()
+  projectId: z.string(),
```
Comment (rajdip-b) [inline apps/platform/src/types/index.ts:45]: Do not use these. We already have type definitions set in the `schema` package.

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: apps/platform/src/components/ui/combobox.tsx [hunk 4]
```diff
@@ -37,19 +38,40 @@ import {
   DialogTrigger
 } from './dialog'
 import { Button } from './button'
+import { WorkspaceSchema } from '@keyshade/schema/schemas'
+
+interface WorkspaceResponse {
```
Comment (rajdip-b) [inline apps/platform/src/components/ui/combobox.tsx:43]: The response object already have the pagination type included. I would highly recommend you to go through the entire schema package once to understand the types.

File: apps/platform/src/components/ui/combobox.tsx [hunk 5]
```diff
@@ -37,19 +38,40 @@ import {
...
+}
 
 async function getAllWorkspace(): Promise<Workspace[] | undefined> {
   try {
-    const workspaceData: Workspace[] =
-      await apiClient.get<Workspace[]>('/workspace')
-
-    const { success, data } = zWorkspace.array().safeParse(workspaceData)
+    const workspaceData: WorkspaceResponse =
+      await apiClient.get<WorkspaceResponse>('/workspace')
```
Comment (rajdip-b) [inline apps/platform/src/components/ui/combobox.tsx:63]: You shouldnt be using direct calls. We have the controllers defined in `api-client`

File: packages/api-client/src/controllers/project.ts [hunk 6]
```diff
@@ -135,6 +135,7 @@ export default class ProjectController {
     request: GetAllProjectsRequest,
     headers: Record<string, string>
   ): Promise<ClientResponse<GetAllProjectsResponse>> {
+
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project.ts:138]: Remove this space

File: apps/platform/src/app/(main)/page.tsx [hunk 7]
```diff
@@ -138,28 +94,37 @@ export default function Index(): JSX.Element {
...
-      .then((data: ProjectWithoutKeys[] | [] | undefined) => {
+    console.log("Current workspace slug: ", currentWorkspace.slug);
+
+    const projectController = new ProjectController(
+          process.env.NEXT_PUBLIC_BACKEND_URL
+    )
+    
+    projectController.getAllProjects({workspaceSlug: currentWorkspace.slug}, {})
+      .then((data: GetAllProjectsResponse) => {
         if (data) {
```
Comment (rajdip-b) [inline apps/platform/src/app/(main)/page.tsx:108]: Thsi isnt the correct pattern. you will get back `data`, `success` and `error`. You would need to check if success is true and then extract it from data. You should use object destructuring in line 107

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: apps/platform/src/components/ui/combobox.tsx (hunk 5 above)
Comment (poswalsameer) [reply apps/platform/src/components/ui/combobox.tsx:63]: > You shouldnt be using direct calls. We have the controllers defined in `api-client`

This was done intentionally, @kriptonian1 was with me. There were some issues from the `api` side on getting the `currentWorkspace`. That's why this was done, just to make sure that the flow atleast works.

File: packages/api-client/src/controllers/project.ts (hunk 6 above)
Comment (poswalsameer) [reply packages/api-client/src/controllers/project.ts:138]: > Remove this space

sure

File: apps/platform/src/app/(main)/page.tsx (hunk 7 above)
Comment (poswalsameer) [reply apps/platform/src/app/(main)/page.tsx:108]: > Thsi isnt the correct pattern. you will get back `data`, `success` and `error`. You would need to check if success is true and then extract it from data. You should use object destructuring in line 107

Cool, I'll do this.

File: apps/platform/src/components/ui/combobox.tsx (hunk 5 above)
Comment (rajdip-b) [reply apps/platform/src/components/ui/combobox.tsx:63]: Oh, but do keep a TODO: in here to change it later on.

**Comment by kriptonian1** [comment]
@rajdip-b if it's working then merge it let's not focus too much on quality

**Comment by rajdip-b** [comment]
> @rajdip-b if it's working then merge it let's not focus too much on quality

I'm focussing on the consistency of things. That's why i've kept it on the hold till the `schema` package gets implemented in here. We don't want to re-do everything. And also, the API is bound to change now that the frontend has started to change.

**Comment by rajdip-b** [comment]
@poswalsameer can you fix the conflicts? Also, i think the type defs are the only thing remaining. once you get that done we can quickly merge this one. its taking a while i must say

**Comment by poswalsameer** [comment]
> @poswalsameer can you fix the conflicts? Also, i think the type defs are the only thing remaining. once you get that done we can quickly merge this one. its taking a while i must say

for the type thing, I had a talk with @kriptonian1, he said that we are going to merge this one, because there are some issues in that file. So yeah, apart from that, everything is done.

**Comment by poswalsameer** [comment]
> @poswalsameer can you fix the conflicts? Also, i think the type defs are the only thing remaining. once you get that done we can quickly merge this one. its taking a while i must say

Fixed the conflicts, for the only change remaining I had a talk with @kriptonian1 about the same, he suggested to merge the changes since there are some issues in that file, so yeah apart from that everything is done from my side.

**Comment by rajdip-b** [comment]
@poswalsameer linting is failing with some GC error. This never happened before. Can you try running this locally and see if the error persists? Also, please update the lockfile.

**Comment by poswalsameer** [comment]
> @poswalsameer linting is failing with some GC error. This never happened before. Can you try running this locally and see if the error persists? Also, please update the lockfile.

Sure, I guess it's happening because of some log statement still remaining in the codefile, let me remove them. Also, can you help me with lockfile, like which things need to be updated there?

**Comment by rajdip-b** [comment]
Delete the pnpm lock file and run pnpm i. that should fix it.

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**